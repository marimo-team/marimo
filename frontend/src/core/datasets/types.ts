/* Copyright 2024 Marimo. All rights reserved. */
import type { DataColumnPreview, DataTable } from "../kernel/messages";

/**
 * A qualified column name, e.g. `table:column`.
 */
export type QualifiedColumn = `${string}:${string}`;

/**
 * A summary of a column's statistics.
 */
export interface ColumnPreviewSummary {
  error?: string;
  total?: number;
  nulls?: number;
  unique?: number;
  min?: number;
  max?: number;
  mean?: number;
  median?: number;
  std?: number;
  p5?: number;
  p25?: number;
  p75?: number;
  p95?: number;
}

export type ColumnPreviewMap = ReadonlyMap<QualifiedColumn, DataColumnPreview>;

/**
 * The datasets state.
 */
export interface DatasetsState {
  tables: DataTable[];
  expandedTables: ReadonlySet<string>;
  expandedColumns: ReadonlySet<QualifiedColumn>;
  columnsPreviews: ColumnPreviewMap;
}
