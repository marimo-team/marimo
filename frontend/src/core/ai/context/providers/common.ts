/* Copyright 2024 Marimo. All rights reserved. */

export const Boosts = {
  LOCAL_TABLE: 5,
  REMOTE_TABLE: 4,
  VARIABLE: 3,
} as const;
