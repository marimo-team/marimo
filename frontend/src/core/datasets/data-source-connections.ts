/* Copyright 2024 Marimo. All rights reserved. */
import { createReducerAndAtoms } from "@/utils/createReducer";
import type {
  DataSourceConnection as DataSourceConnectionType,
  DataTable,
  SQLTablePreview,
} from "../kernel/messages";
import type { TypedString } from "@/utils/typed";
import type { VariableName } from "../variables/types";
import { atom } from "jotai";
import { store } from "../state/jotai";
import { datasetTablesAtom } from "./state";
import { Logger } from "@/utils/Logger";

export type ConnectionName = TypedString<"ConnectionName">;

export const DEFAULT_ENGINE = "__marimo_duckdb" as ConnectionName;

// Extend the backend type but override the name property with the strongly typed ConnectionName
export interface DataSourceConnection
  extends Omit<DataSourceConnectionType, "name"> {
  name: ConnectionName;
}

export interface DataSourceState {
  latestEngineSelected: ConnectionName;
  connectionsMap: ReadonlyMap<ConnectionName, DataSourceConnection>;
}

function initialState(): DataSourceState {
  return {
    latestEngineSelected: DEFAULT_ENGINE,
    connectionsMap: new Map().set(DEFAULT_ENGINE, {
      name: DEFAULT_ENGINE,
      dialect: "duckdb",
      source: "duckdb",
      display_name: "DuckDB In-Memory",
      databases: [],
    }),
  };
}

const {
  reducer,
  createActions,
  valueAtom: dataSourceConnectionsAtom,
  useActions: useDataSourceActions,
} = createReducerAndAtoms(initialState, {
  addDataSourceConnection: (
    state: DataSourceState,
    opts: { connections: DataSourceConnection[] },
  ): DataSourceState => {
    if (opts.connections.length === 0) {
      return state;
    }

    const { latestEngineSelected, connectionsMap } = state;

    // update existing connections with latest values
    // add new ones if they don't exist
    // Backend will dedupe by connection name & keep the latest, so we use this as the key
    const newMap = new Map(connectionsMap);
    for (const conn of opts.connections) {
      newMap.set(conn.name, conn);
    }

    return {
      latestEngineSelected,
      connectionsMap: newMap,
    };
  },

  // Keep default engine and any connections that are used by variables
  filterDataSourcesFromVariables: (
    state: DataSourceState,
    variableNames: VariableName[],
  ) => {
    const { latestEngineSelected, connectionsMap } = state;
    const names = new Set(variableNames);
    const newMap = new Map(
      [...connectionsMap].filter(([name]) => {
        if (name === DEFAULT_ENGINE) {
          return true;
        }
        return names.has(name as unknown as VariableName);
      }),
    );
    return {
      // If the latest engine selected is not in the new map, use the default engine
      latestEngineSelected: newMap.has(latestEngineSelected)
        ? latestEngineSelected
        : DEFAULT_ENGINE,
      connectionsMap: newMap,
    };
  },

  clearDataSourceConnections: (): DataSourceState => ({
    latestEngineSelected: DEFAULT_ENGINE,
    connectionsMap: new Map(),
  }),

  removeDataSourceConnection: (
    state: DataSourceState,
    connectionName: ConnectionName,
  ): DataSourceState => {
    const { latestEngineSelected, connectionsMap } = state;

    const newMap = new Map(connectionsMap);
    newMap.delete(connectionName);
    return {
      latestEngineSelected: newMap.has(latestEngineSelected)
        ? latestEngineSelected
        : DEFAULT_ENGINE,
      connectionsMap: newMap,
    };
  },
});

export { dataSourceConnectionsAtom, useDataSourceActions };

export const dataConnectionsMapAtom = atom(
  (get) => get(dataSourceConnectionsAtom).connectionsMap,
);

export function setLatestEngineSelected(engine: ConnectionName) {
  const existing = store.get(dataSourceConnectionsAtom);
  // Don't update the map if the engine is not in the map
  if (existing.connectionsMap.has(engine)) {
    store.set(dataSourceConnectionsAtom, {
      ...existing,
      latestEngineSelected: engine,
    });
  }
}

export const exportedForTesting = {
  reducer,
  createActions,
  initialState,
};

// Hook to get & persist SQL table previews
// Acts as a cache
export const tablePreviewsAtom = atom<ReadonlyMap<string, SQLTablePreview>>(
  new Map<string, SQLTablePreview>(),
);

// If you need to get table names from all connections & local datasets, use this atom
// When a table name is used in multiple connections, we need to use a more qualified name
export const allTablesAtom = atom((get) => {
  const datasets = store.get(datasetTablesAtom);
  const connections = get(dataSourceConnectionsAtom).connectionsMap;
  const tableNames = new Map<string, DataTable>();

  for (const dataset of datasets) {
    tableNames.set(dataset.name, dataset);
  }

  for (const conn of connections.values()) {
    for (const database of conn.databases) {
      // If there is only one database, it is the default
      const isDefaultDb =
        database.name === conn.default_database || conn.databases.length === 1;

      for (const schema of database.schemas) {
        const isDefaultSchema = schema.name === conn.default_schema;

        for (const table of schema.tables) {
          let nameToSave: string;

          // If the database and schema are default, we can use the table name directly
          // Otherwise, we need to qualify the table name
          // We also need to use the more qualified name if there are collisions
          nameToSave = table.name;

          if (isDefaultDb && isDefaultSchema && !tableNames.has(nameToSave)) {
            tableNames.set(nameToSave, table);
            continue;
          }

          nameToSave = `${schema.name}.${table.name}`;

          if (isDefaultDb && !tableNames.has(nameToSave)) {
            tableNames.set(nameToSave, table);
            continue;
          }

          nameToSave = `${database.name}.${schema.name}.${table.name}`;

          if (tableNames.has(nameToSave)) {
            Logger.warn(`Table name collision for ${nameToSave}. Skipping.`);
          } else {
            tableNames.set(nameToSave, table);
          }
        }
      }
    }
  }

  return tableNames;
});
