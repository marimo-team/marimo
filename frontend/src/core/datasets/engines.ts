/* Copyright 2024 Marimo. All rights reserved. */
import type { TypedString } from "@/utils/typed";

export type ConnectionName = TypedString<"ConnectionName">;

// DuckDB engine is treated as the default engine
// As it doesn't require passing an engine variable to the backend
// Keep this in sync with the backend name
export const DUCKDB_ENGINE = "__marimo_duckdb" as ConnectionName;
export const INTERNAL_SQL_ENGINES = new Set([DUCKDB_ENGINE]);
