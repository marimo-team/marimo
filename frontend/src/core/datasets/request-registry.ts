/* Copyright 2024 Marimo. All rights reserved. */
import type { SQLTablePreview, SQLTableListPreview } from "../kernel/messages";
import { DeferredRequestRegistry } from "../network/DeferredRequestRegistry";
import { previewSQLTable, previewSQLTableList } from "../network/requests";
import type {
  PreviewSQLTableRequest,
  PreviewSQLTableListRequest,
} from "../network/types";

// We make a request to the backend to preview the table, passing in Engine, DB, Schema, and Table
// The backend returns data tables, which could also exist in other engines, dbs, schemas
// Thus, we use the request ID pattern to match the response to the request

export const PreviewSQLTable = new DeferredRequestRegistry<
  Omit<PreviewSQLTableRequest, "requestId">,
  SQLTablePreview
>("sql-table-preview", async (requestId, req) => {
  await previewSQLTable({
    requestId: requestId,
    ...req,
  });
});

export const PreviewSQLTableList = new DeferredRequestRegistry<
  Omit<PreviewSQLTableListRequest, "requestId">,
  SQLTableListPreview
>("sql-table-list-preview", async (requestId, req) => {
  await previewSQLTableList({
    requestId: requestId,
    ...req,
  });
});
