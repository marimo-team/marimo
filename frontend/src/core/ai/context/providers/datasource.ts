/* Copyright 2024 Marimo. All rights reserved. */

import type { Completion } from "@codemirror/autocomplete";
import { createRoot } from "react-dom/client";
import { dbDisplayName } from "@/components/databases/display";
import { renderDatasourceInfo } from "@/core/codemirror/language/languages/sql/renderers";
import {
  type ConnectionsMap,
  type DatasetTablesMap,
  getTableType,
} from "@/core/datasets/data-source-connections";
import {
  type ConnectionName,
  INTERNAL_SQL_ENGINES,
} from "@/core/datasets/engines";
import type { DataSourceConnection, DataTable } from "@/core/kernel/messages";
import type { AIContextItem } from "../registry";
import { AIContextProvider } from "../registry";
import { contextToXml } from "../utils";
import { Boosts, Sections } from "./common";

type NamedDatasource = Omit<
  DataSourceConnection,
  "name" | "display_name" | "source"
> & {
  // Easier for the AI to write mo.sql with engine name
  engine_name?: ConnectionName;
};

export interface DatasourceContextItem extends AIContextItem {
  type: "datasource";
  // For internal engine, it can have both connection and data tables
  // For external engines, the data is just a DataSourceConnection
  data: {
    connection: DataSourceConnection;
    tables?: DataTable[];
  };
}

export class DatasourceContextProvider extends AIContextProvider<DatasourceContextItem> {
  readonly title = "Datasource";
  readonly mentionPrefix = "@";
  readonly contextType = "datasource";
  private connectionsMap: ConnectionsMap;
  private dataframes: DataTable[];

  constructor(connectionsMap: ConnectionsMap, tablesMap: DatasetTablesMap) {
    super();
    this.connectionsMap = connectionsMap;
    this.dataframes = [...tablesMap.values()].filter(
      (table: DataTable) => getTableType(table) === "dataframe",
    );
  }

  getItems(): DatasourceContextItem[] {
    return [...this.connectionsMap.values()]
      .map((connection): DatasourceContextItem | null => {
        let description = "Database schema.";
        const data: DatasourceContextItem["data"] = {
          connection: connection,
        };

        if (INTERNAL_SQL_ENGINES.has(connection.name)) {
          data.tables = this.dataframes;
          description =
            "Database schema and the dataframes that can be queried";
        }

        // Hide empty datasources
        const hasNoTables =
          connection.databases.length === 0 && (data.tables?.length ?? 0) === 0;
        if (hasNoTables) {
          return null;
        }

        return {
          uri: this.asURI(connection.name),
          name: connection.name,
          description: description,
          type: this.contextType,
          data: data,
        };
      })
      .filter(Boolean);
  }

  formatContext(item: DatasourceContextItem): string {
    const data = item.data;
    // Remove certain fields that are not needed in the context
    const { name, display_name, source, ...filteredDatasource } =
      data.connection;

    let datasource = filteredDatasource;
    const isInternalEngine = INTERNAL_SQL_ENGINES.has(name as ConnectionName);
    if (!isInternalEngine) {
      const namedDatasource: NamedDatasource = {
        ...filteredDatasource,
        engine_name: name as ConnectionName, // Add the engine name for external engines
      };
      datasource = namedDatasource;
    }

    return contextToXml({
      type: this.contextType,
      data: {
        connection: datasource,
        tables: data.tables,
      },
    });
  }

  formatCompletion(item: DatasourceContextItem): Completion {
    const datasource = item.data;

    const dataConnection = datasource.connection;
    const dataframes = datasource.tables;

    let label = dataConnection.name;
    if (INTERNAL_SQL_ENGINES.has(dataConnection.name as ConnectionName)) {
      label = "In-Memory";
    }

    return {
      label: `@${label}`,
      displayLabel: label,
      detail: dbDisplayName(dataConnection.dialect),
      boost: Boosts.MEDIUM,
      type: this.contextType,
      section: Sections.DATA_SOURCES,
      info: () => {
        const infoContainer = document.createElement("div");
        infoContainer.classList.add("mo-cm-tooltip", "docs-documentation");

        // Use React to render the datasource info
        const root = createRoot(infoContainer);
        root.render(renderDatasourceInfo(dataConnection, dataframes));

        return infoContainer;
      },
    };
  }
}
