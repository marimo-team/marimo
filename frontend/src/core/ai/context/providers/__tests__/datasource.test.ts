/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { beforeEach, describe, expect, it } from "vitest";
import type { CellId } from "@/core/cells/ids";
import type {
  ConnectionsMap,
  DatasetTablesMap,
} from "@/core/datasets/data-source-connections";
import { DUCKDB_ENGINE } from "@/core/datasets/engines";
import type { DataSourceConnection, DataTable } from "@/core/kernel/messages";
import { Boosts, Sections } from "../common";
import { DatasourceContextProvider, getDatasourceContext } from "../datasource";

// Mock data for testing
const createMockDataSourceConnection = (
  name: string,
  options: Partial<DataSourceConnection> = {},
): DataSourceConnection => ({
  name,
  dialect: "duckdb",
  source: "duckdb",
  display_name: `Test ${name}`,
  databases: [
    {
      name: "main",
      dialect: "duckdb",
      schemas: [
        {
          name: "public",
          tables: [
            {
              name: "users",
              source_type: "connection",
              source: name,
              num_rows: 100,
              num_columns: 3,
              variable_name: null,
              columns: [],
            },
            {
              name: "orders",
              source_type: "connection",
              source: name,
              num_rows: 50,
              num_columns: 4,
              variable_name: null,
              columns: [],
            },
          ],
        },
        {
          name: "analytics",
          tables: [
            {
              name: "events",
              source_type: "connection",
              source: name,
              num_rows: 200,
              num_columns: 5,
              variable_name: null,
              columns: [],
            },
          ],
        },
      ],
    },
  ],
  ...options,
});

const createMockConnectionsMap = (
  connections: DataSourceConnection[],
): ConnectionsMap => {
  const map = new Map();
  connections.forEach((conn) => {
    map.set(conn.name, conn);
  });
  return map;
};

const createMockDataTable = (
  name: string,
  options: Partial<DataTable> = {},
): DataTable => ({
  name,
  source: "local",
  source_type: "local",
  num_rows: 100,
  num_columns: 3,
  variable_name: name, // This makes it a dataframe
  columns: [
    {
      name: "id",
      type: "integer",
      external_type: "INTEGER",
      sample_values: [1, 2, 3],
    },
    {
      name: "name",
      type: "string",
      external_type: "VARCHAR",
      sample_values: ["Alice", "Bob", "Charlie"],
    },
    {
      name: "age",
      type: "integer",
      external_type: "INTEGER",
      sample_values: [25, 30, 35],
    },
  ],
  engine: null,
  indexes: null,
  primary_keys: null,
  type: "table",
  ...options,
});

const createMockTablesMap = (tables: DataTable[]): DatasetTablesMap => {
  const map = new Map();
  tables.forEach((table) => {
    map.set(table.name, table);
  });
  return map;
};

describe("DatasourceContextProvider", () => {
  let provider: DatasourceContextProvider;
  let connectionsMap: ConnectionsMap;
  let tablesMap: DatasetTablesMap;

  beforeEach(() => {
    connectionsMap = createMockConnectionsMap([
      createMockDataSourceConnection(DUCKDB_ENGINE),
      createMockDataSourceConnection("postgres", {
        dialect: "postgresql",
        source: "postgresql",
        display_name: "PostgreSQL Database",
        databases: [
          {
            name: "production",
            dialect: "postgresql",
            schemas: [
              {
                name: "public",
                tables: [
                  {
                    name: "customers",
                    source_type: "connection",
                    source: "postgres",
                    num_rows: 1000,
                    num_columns: 8,
                    variable_name: null,
                    columns: [],
                  },
                  {
                    name: "products",
                    source_type: "connection",
                    source: "postgres",
                    num_rows: 500,
                    num_columns: 6,
                    variable_name: null,
                    columns: [],
                  },
                  {
                    name: "sales",
                    source_type: "connection",
                    source: "postgres",
                    num_rows: 5000,
                    num_columns: 10,
                    variable_name: null,
                    columns: [],
                  },
                ],
              },
            ],
          },
        ],
      }),
    ]);
    tablesMap = createMockTablesMap([
      createMockDataTable("users"),
      createMockDataTable("orders"),
      createMockDataTable("events"),
    ]);
    provider = new DatasourceContextProvider(connectionsMap, tablesMap);
  });

  describe("provider properties", () => {
    it("should have correct provider properties", () => {
      expect(provider.title).toBe("Datasource");
      expect(provider.mentionPrefix).toBe("@");
      expect(provider.contextType).toBe("datasource");
    });
  });

  describe("getItems", () => {
    it("should return empty array when no connections", () => {
      const emptyMap = createMockConnectionsMap([]);
      const emptyProvider = new DatasourceContextProvider(emptyMap, new Map());
      const items = emptyProvider.getItems();
      expect(items).toEqual([]);
    });

    it("should return datasource items when connections exist", () => {
      const items = provider.getItems();

      expect(items).toHaveLength(2);

      // Check first item (duckdb)
      expect(items[0]).toMatchObject({
        name: DUCKDB_ENGINE,
        type: "datasource",
        data: {
          connection: {
            name: DUCKDB_ENGINE,
            dialect: "duckdb",
            source: "duckdb",
            display_name: `Test ${DUCKDB_ENGINE}`,
          },
        },
      });

      // Check second item (postgres)
      expect(items[1]).toMatchObject({
        name: "postgres",
        type: "datasource",
        data: {
          connection: {
            name: "postgres",
            dialect: "postgresql",
            source: "postgresql",
            display_name: "PostgreSQL Database",
          },
        },
      });

      // Check URIs are properly formatted
      expect(items[0].uri).toBe(`datasource://${DUCKDB_ENGINE}`);
      expect(items[1].uri).toBe("datasource://postgres");
    });

    it("should include dataframes for internal SQL engines", () => {
      const items = provider.getItems();
      const duckdbItem = items.find((item) => item.name === DUCKDB_ENGINE)!;

      // DuckDB is an internal engine, so it should have tables (dataframes)
      expect(duckdbItem.data.tables).toBeDefined();
      expect(duckdbItem.data.tables).toHaveLength(3);
      expect(duckdbItem.data.tables?.map((t) => t.name)).toEqual([
        "users",
        "orders",
        "events",
      ]);

      // PostgreSQL is external, so it should not have tables
      const postgresItem = items.find((item) => item.name === "postgres")!;
      expect(postgresItem.data.tables).toBeUndefined();
    });

    it("should handle connections with no databases", () => {
      const emptyDbConnection = createMockDataSourceConnection("empty", {
        databases: [],
      });
      const mapWithEmpty = createMockConnectionsMap([emptyDbConnection]);
      const providerWithEmpty = new DatasourceContextProvider(
        mapWithEmpty,
        new Map(),
      );

      const items = providerWithEmpty.getItems();
      expect(items).toHaveLength(0);
    });

    it("should handle connections with databases but no schemas", () => {
      const emptySchemaConnection = createMockDataSourceConnection(
        "empty-schema",
        {
          databases: [
            {
              name: "empty_db",
              dialect: "duckdb",
              schemas: [],
            },
          ],
        },
      );
      const mapWithEmptySchema = createMockConnectionsMap([
        emptySchemaConnection,
      ]);
      const providerWithEmptySchema = new DatasourceContextProvider(
        mapWithEmptySchema,
        new Map(),
      );

      const items = providerWithEmptySchema.getItems();
      expect(items).toHaveLength(1);
      expect(items[0].data.connection.databases[0].schemas).toEqual([]);
    });
  });

  describe("formatCompletion", () => {
    it("should format completion for datasource with tables", () => {
      const items = provider.getItems();
      const duckdbItem = items.find((item) => item.name === DUCKDB_ENGINE)!;
      const completion = provider.formatCompletion(duckdbItem);

      expect(completion).toMatchObject({
        label: "@In-Memory",
        displayLabel: "In-Memory",
        detail: "DuckDB",
        boost: Boosts.MEDIUM,
        type: "datasource",
        section: Sections.DATA_SOURCES,
      });

      expect(completion.info).toBeDefined();
    });

    it("should format completion for datasource with no tables", () => {
      const emptyConnection = createMockDataSourceConnection("empty", {
        databases: [
          {
            name: "empty_db",
            dialect: "duckdb",
            schemas: [
              {
                name: "empty_schema",
                tables: [],
              },
            ],
          },
        ],
      });
      const mapWithEmpty = createMockConnectionsMap([emptyConnection]);
      const providerWithEmpty = new DatasourceContextProvider(
        mapWithEmpty,
        new Map(),
      );

      const items = providerWithEmpty.getItems();
      const completion = provider.formatCompletion(items[0]);

      expect(completion.detail).toBe("DuckDB");
    });

    it("should format completion for postgres connection", () => {
      const items = provider.getItems();
      const postgresItem = items.find((item) => item.name === "postgres")!;
      const completion = provider.formatCompletion(postgresItem);

      expect(completion).toMatchObject({
        label: "@postgres",
        displayLabel: "postgres",
        detail: "PostgreSQL",
        boost: Boosts.MEDIUM,
        type: "datasource",
        section: Sections.DATA_SOURCES,
      });
    });

    it("should format completion for in-memory engine", () => {
      const duckdbConnection = createMockDataSourceConnection(DUCKDB_ENGINE, {
        dialect: "duckdb",
        source: "duckdb",
        display_name: "DuckDB (In-Memory)",
      });
      const mapWithInMemory = createMockConnectionsMap([duckdbConnection]);
      const providerWithInMemory = new DatasourceContextProvider(
        mapWithInMemory,
        new Map(),
      );
      const items = providerWithInMemory.getItems();
      const inMemoryItem = items.find((item) => item.name === DUCKDB_ENGINE)!;
      const completion = provider.formatCompletion(inMemoryItem);

      expect(completion).toMatchObject({
        label: "@In-Memory",
        displayLabel: "In-Memory",
        detail: "DuckDB",
        boost: Boosts.MEDIUM,
        type: "datasource",
        section: Sections.DATA_SOURCES,
      });
    });
  });

  describe("formatContext", () => {
    it("should format context for internal duckdb with multiple tables", () => {
      const items = provider.getItems();
      const duckdbItem = items.find((item) => item.name === DUCKDB_ENGINE)!;
      const context = provider.formatContext(duckdbItem);

      expect(context).not.toContain('"engine_name":"__marimo_duckdb"');
      expect(context).toMatchSnapshot("internal-datasource-context");
    });

    it("should include dataframes in context for internal engines", () => {
      const items = provider.getItems();
      const duckdbItem = items.find((item) => item.name === DUCKDB_ENGINE)!;
      const context = provider.formatContext(duckdbItem);

      // Should include the dataframes in the context
      expect(context).toContain('"name":"users"');
      expect(context).toContain('"name":"orders"');
      expect(context).toContain('"name":"events"');
      expect(context).toContain('"variable_name":"users"');
      expect(context).toContain('"variable_name":"orders"');
      expect(context).toContain('"variable_name":"events"');
    });

    it("should format context for postgres datasource", () => {
      const items = provider.getItems();
      const postgresItem = items.find((item) => item.name === "postgres")!;
      const context = provider.formatContext(postgresItem);

      expect(context).toContain('"engine_name":"postgres"');
      expect(context).toMatchSnapshot("postgres-datasource-context");
    });

    it("should format context for datasource with no tables", () => {
      const emptyConnection = createMockDataSourceConnection("empty", {
        databases: [
          {
            name: "empty_db",
            dialect: "duckdb",
            schemas: [
              {
                name: "empty_schema",
                tables: [],
              },
            ],
          },
        ],
      });
      const mapWithEmpty = createMockConnectionsMap([emptyConnection]);
      const providerWithEmpty = new DatasourceContextProvider(
        mapWithEmpty,
        new Map(),
      );

      const items = providerWithEmpty.getItems();
      const context = providerWithEmpty.formatContext(items[0]);

      expect(context).toContain('"dialect":"duckdb"');
      expect(context).not.toContain('"name":"Test empty"');
      expect(context).not.toContain('"source":"duckdb"');
      expect(context).not.toContain('"display_name":"Test empty"');
    });

    it("should format context for datasource with multiple databases", () => {
      const multiDbConnection = createMockDataSourceConnection("multi", {
        default_database: "db1",
        default_schema: "schema1",
        databases: [
          {
            name: "db1",
            dialect: "duckdb",
            schemas: [
              {
                name: "schema1",
                tables: [
                  {
                    name: "table1",
                    source_type: "connection",
                    source: "multi",
                    num_rows: 10,
                    num_columns: 2,
                    variable_name: null,
                    columns: [],
                  },
                  {
                    name: "table2",
                    source_type: "connection",
                    source: "multi",
                    num_rows: 20,
                    num_columns: 3,
                    variable_name: null,
                    columns: [],
                  },
                ],
              },
            ],
          },
          {
            name: "db2",
            dialect: "duckdb",
            schemas: [
              {
                name: "schema2",
                tables: [
                  {
                    name: "table3",
                    source_type: "connection",
                    source: "multi",
                    num_rows: 15,
                    num_columns: 4,
                    variable_name: null,
                    columns: [],
                  },
                ],
              },
              {
                name: "schema3",
                tables: [],
              },
            ],
          },
        ],
      });
      const mapWithMulti = createMockConnectionsMap([multiDbConnection]);
      const providerWithMulti = new DatasourceContextProvider(
        mapWithMulti,
        new Map(),
      );

      const items = providerWithMulti.getItems();
      const context = provider.formatContext(items[0]);

      expect(context).toContain('"dialect":"duckdb"');
      expect(context).toContain('"default_database":"db1"');
      expect(context).toContain('"default_schema":"schema1"');
      expect(context).toMatchSnapshot("multi-db-context");
    });
  });

  describe("edge cases", () => {
    it("should highlight default database and schema", () => {
      const connectionWithDefaults = createMockDataSourceConnection("test", {
        default_database: "main",
        default_schema: "public",
        databases: [
          {
            name: "main",
            dialect: "duckdb",
            schemas: [
              {
                name: "public",
                tables: [
                  {
                    name: "users",
                    source_type: "connection",
                    source: "test",
                    num_rows: 100,
                    num_columns: 3,
                    variable_name: null,
                    columns: [],
                  },
                ],
              },
              {
                name: "analytics",
                tables: [],
              },
            ],
          },
        ],
      });
      const mapWithDefaults = createMockConnectionsMap([
        connectionWithDefaults,
      ]);
      const providerWithDefaults = new DatasourceContextProvider(
        mapWithDefaults,
        new Map(),
      );

      const items = providerWithDefaults.getItems();
      const completion = providerWithDefaults.formatCompletion(items[0]);

      expect(completion.info).toBeDefined();
      expect(typeof completion.info).toBe("function");
    });

    it("should handle connections with special characters in names", () => {
      const specialConnection = createMockDataSourceConnection("test-db_123", {
        display_name: "Test DB (123)",
      });
      const mapWithSpecial = createMockConnectionsMap([specialConnection]);
      const providerWithSpecial = new DatasourceContextProvider(
        mapWithSpecial,
        new Map(),
      );

      const items = providerWithSpecial.getItems();
      expect(items).toHaveLength(1);
      expect(items[0].name).toBe("test-db_123");
      expect(items[0].uri).toBe("datasource://test-db_123");

      const context = providerWithSpecial.formatContext(items[0]);
      expect(context).toContain('"dialect":"duckdb"');
    });

    it("should handle very large numbers of tables", () => {
      const largeConnection = createMockDataSourceConnection("large", {
        databases: [
          {
            name: "large_db",
            dialect: "duckdb",
            schemas: [
              {
                name: "large_schema",
                tables: Array.from({ length: 100 }, (_, i) => ({
                  name: `table_${i}`,
                  source_type: "connection" as const,
                  source: "large",
                  num_rows: i + 1,
                  num_columns: 2,
                  variable_name: null,
                  columns: [],
                })),
              },
            ],
          },
        ],
      });
      const mapWithLarge = createMockConnectionsMap([largeConnection]);
      const providerWithLarge = new DatasourceContextProvider(
        mapWithLarge,
        new Map(),
      );

      const items = providerWithLarge.getItems();
      const completion = providerWithLarge.formatCompletion(items[0]);
      expect(completion.detail).toBe("DuckDB");

      const context = providerWithLarge.formatContext(items[0]);
      // Since we now return the entire data structure, check for basic properties
      expect(context).not.toContain('"name":"Test large"');
      expect(context).toContain('"dialect":"duckdb"');
      expect(context).not.toContain('"source":"duckdb"');
    });
  });
});

describe("getDatasourceContext", () => {
  it("should return null if no cell ID is found", () => {
    const context = getDatasourceContext("1" as CellId);
    expect(context).toBeNull();
  });
});
