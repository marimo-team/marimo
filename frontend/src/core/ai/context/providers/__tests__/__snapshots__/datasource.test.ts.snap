// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatasourceContextProvider > formatContext > should format context for datasource with multiple databases > multi-db-context 1`] = `"<datasource connection="{"dialect":"duckdb","databases":[{"name":"db1","dialect":"duckdb","schemas":[{"name":"schema1","tables":[{"name":"table1","source_type":"connection","source":"multi","num_rows":10,"num_columns":2,"variable_name":null,"columns":[]},{"name":"table2","source_type":"connection","source":"multi","num_rows":20,"num_columns":3,"variable_name":null,"columns":[]}]}]},{"name":"db2","dialect":"duckdb","schemas":[{"name":"schema2","tables":[{"name":"table3","source_type":"connection","source":"multi","num_rows":15,"num_columns":4,"variable_name":null,"columns":[]}]},{"name":"schema3","tables":[]}]}],"default_database":"db1","default_schema":"schema1","engine_name":"multi"}"></datasource>"`;

exports[`DatasourceContextProvider > formatContext > should format context for internal duckdb with multiple tables > internal-datasource-context 1`] = `"<datasource connection="{"dialect":"duckdb","databases":[{"name":"main","dialect":"duckdb","schemas":[{"name":"public","tables":[{"name":"users","source_type":"connection","source":"__marimo_duckdb","num_rows":100,"num_columns":3,"variable_name":null,"columns":[]},{"name":"orders","source_type":"connection","source":"__marimo_duckdb","num_rows":50,"num_columns":4,"variable_name":null,"columns":[]}]},{"name":"analytics","tables":[{"name":"events","source_type":"connection","source":"__marimo_duckdb","num_rows":200,"num_columns":5,"variable_name":null,"columns":[]}]}]}]}" tables="[{"name":"users","source":"local","source_type":"local","num_rows":100,"num_columns":3,"variable_name":"users","columns":[{"name":"id","type":"integer","external_type":"INTEGER","sample_values":[1,2,3]},{"name":"name","type":"string","external_type":"VARCHAR","sample_values":["Alice","Bob","Charlie"]},{"name":"age","type":"integer","external_type":"INTEGER","sample_values":[25,30,35]}],"engine":null,"indexes":null,"primary_keys":null,"type":"table"},{"name":"orders","source":"local","source_type":"local","num_rows":100,"num_columns":3,"variable_name":"orders","columns":[{"name":"id","type":"integer","external_type":"INTEGER","sample_values":[1,2,3]},{"name":"name","type":"string","external_type":"VARCHAR","sample_values":["Alice","Bob","Charlie"]},{"name":"age","type":"integer","external_type":"INTEGER","sample_values":[25,30,35]}],"engine":null,"indexes":null,"primary_keys":null,"type":"table"},{"name":"events","source":"local","source_type":"local","num_rows":100,"num_columns":3,"variable_name":"events","columns":[{"name":"id","type":"integer","external_type":"INTEGER","sample_values":[1,2,3]},{"name":"name","type":"string","external_type":"VARCHAR","sample_values":["Alice","Bob","Charlie"]},{"name":"age","type":"integer","external_type":"INTEGER","sample_values":[25,30,35]}],"engine":null,"indexes":null,"primary_keys":null,"type":"table"}]"></datasource>"`;

exports[`DatasourceContextProvider > formatContext > should format context for postgres datasource > postgres-datasource-context 1`] = `"<datasource connection="{"dialect":"postgresql","databases":[{"name":"production","dialect":"postgresql","schemas":[{"name":"public","tables":[{"name":"customers","source_type":"connection","source":"postgres","num_rows":1000,"num_columns":8,"variable_name":null,"columns":[]},{"name":"products","source_type":"connection","source":"postgres","num_rows":500,"num_columns":6,"variable_name":null,"columns":[]},{"name":"sales","source_type":"connection","source":"postgres","num_rows":5000,"num_columns":10,"variable_name":null,"columns":[]}]}]}],"engine_name":"postgres"}"></datasource>"`;
